File: generate_seed.py

import re
import json

def generate_sql():
    with open('data/algeria-locations.ts', 'r', encoding='utf-8') as f:
        content = f.read()

    # Find the array content
    array_match = re.search(r'algeriaLocations: WilayaData\[\] = (\[.*\]);', content, re.DOTALL)
    if not array_match:
        print("Could not find algeriaLocations array")
        return

    json_str = array_match.group(1)
    # Basic cleanup to make it JSON-ish
    json_str = re.sub(r'//.*', '', json_str)
    
    try:
        data = json.loads(json_str)
    except Exception as e:
        # If it's not perfect JSON (e.g. trailing commas, missing quotes), try a looser approach
        print(f"JSON load failed, attempting regex extraction: {e}")
        data = []
        wilaya_blocks = re.findall(r'\{\s*"id":\s*"(\d+)",\s*"name":\s*"([^"]+)",\s*"communes":\s*\[(.*?)\]\s*\}', json_str, re.DOTALL)
        for w_id, w_name, c_text in wilaya_blocks:
            communes = re.findall(r'"([^"]+)"', c_text)
            data.append({"id": w_id, "name": w_name, "communes": communes})

    sql = "-- Full Algerian Communes Seed\n"
    sql += "TRUNCATE TABLE communes CASCADE;\n"
    sql += "INSERT INTO communes (wilaya_id, name) VALUES\n"
    
    values = []
    for wilaya in data:
        w_id = int(wilaya['id'])
        for commune in wilaya['communes']:
            escaped_name = commune.replace("'", "''")
            values.append(f"({w_id}, '{escaped_name}')")
            
    sql += ",\n".join(values) + ";\n"
    
    with open('full_communes_seed.sql', 'w', encoding='utf-8') as f:
        f.write(sql)
    print(f"Generated seed for {len(values)} communes")

if __name__ == "__main__":
    generate_sql()

File: generate_seed.js

const fs = require('fs');

const content = fs.readFileSync('data/algeria-locations.ts', 'utf-8');

// Use regex to extract the array content
// The array starts after 'algeriaLocations: WilayaData[] = '
const startIndex = content.indexOf('[');
const endIndex = content.lastIndexOf(']') + 1;
const jsonLike = content.slice(startIndex, endIndex);

// A very loose parser for the specific format in algeria-locations.ts
function parseLocations(text) {
    const results = [];
    // Wilaya block regex: { "id": "...", "name": "...", "communes": [...] }
    const wilayaRegex = /\{\s*"id":\s*"(\d+)",\s*"name":\s*"([^"]+)",\s*"communes":\s*\[(.*?)\]\s*\}/gs;
    
    let match;
    while ((match = wilayaRegex.exec(text)) !== null) {
        const id = match[1];
        const name = match[2];
        const communesText = match[3];
        // Extract communes names between quotes
        const communeNames = [];
        const communeRegex = /"([^"]+)"/g;
        let cMatch;
        while ((cMatch = communeRegex.exec(communesText)) !== null) {
            communeNames.push(cMatch[1]);
        }
        results.push({ id, name, communes: communeNames });
    }
    return results;
}

const locations = parseLocations(jsonLike);
console.log(`Found ${locations.length} wilayas.`);

let sql = "-- Full Algerian Communes Seed\n";
sql += "TRUNCATE TABLE communes CASCADE;\n";
sql += "INSERT INTO communes (wilaya_id, name) VALUES\n";

const values = [];
locations.forEach(wilaya => {
    const wId = parseInt(wilaya.id);
    wilaya.communes.forEach(commune => {
        const escapedName = commune.replace(/'/g, "''");
        values.push(`(${wId}, '${escapedName}')`);
    });
});

sql += values.join(',\n') + ';\n';

fs.writeFileSync('full_communes_seed.sql', sql);
console.log(`Generated seed for ${values.length} communes in full_communes_seed.sql`);
